extensions/amp-onetap-google/0.1/amp-onetap-google.js

```diff
@@ -0,0 +1,322 @@
+import {Layout} from '../../../src/layout';
+import {
+  assertDoesNotContainDisplay,
+  setStyle,
+  setStyles,
+  toggle,
+} from '../../../src/style';
+import {isExperimentOn} from '../../../src/experiments';
+import {userAssert} from '../../../src/log';
+
+/** @const {string} */
+const TAG = 'amp-onetap-google';
+
+/** @const {string} */
+const IFRAME_ID = 'amp-onetap-google-intermediate-iframe';
+
+/** @const {Object} */
+const ACTIONS = {
+  RESIZE: 'onetap_iframe_resize',
+  SKIP: 'onetap_iframe_close',
+  DONE: 'onetap_iframe_check_status',
+  SET_UI_MODE: 'onetap_iframe_ui_mode',
+  SET_TAP_OUTSIDE_MODE: 'onetap_iframe_tap_outside_mode',
+};
+
+/** @const {Object} */
+const UI_MODES = {
+  BOTTOM_SHEET: 'bottomsheet',
+  CARD: 'card',
+};
+
+/** @const {Object} */
+const cardIframeStyle = {
+  border: 'none',
+  margin: 0,
+  padding: 0,
+  position: 'fixed',
+  right: '0',
+  left: 'auto',
+  top: '0',
+  bottom: 'auto',
+};
+
+/** @const {Object} */
+const bottomsheetIframeStyle = {
+  border: 'none',
+  margin: 0,
+  padding: 0,
+  position: 'fixed',
+  right: 'auto',
+  left: '0',
+  top: 'auto',
+  bottom: '0',
+};
+
+/** @const {Object} */
+const elementStyle = {
+  height: '0',
+  width: '0',
+  'z-index': 9999,
+  border: 'none',
+  position: 'fixed',
+  left: '0',
+  bottom: '0',
+};
+
+export class AmpOnetapGoogle extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    this.checkExperimentEnabled_();
+
+    /** @private @const {string} */
+    this.intermediateIframeUrl_ = this.element.getAttribute('data-iframe_url');
+
+    /** @private @const {string} */
+    this.intermediateIframeOrigin_ = new URL(
+      this.intermediateIframeUrl_
+    ).origin;
+
+    /** @private {!Element} */
+    this.doc_ = this.element.ownerDocument;
+
+    /** @private {boolean=} */
+    this.isBottomSheetUiMode_ = undefined;
+
+    /** @private {!Element=} */
+    this.intermediateIframe_;
+
+    /** @private {boolean} */
+    this.isIntermediateIframeVisible_ = false;
+
+    /** @private {boolean} */
+    this.cancelOnTapOutside_ = false;
+
+    /** @private */
+    this.onClickOutside_ = () => {
+      if (this.cancelOnTapOutside_ && this.isIntermediateIframeVisible_) {
+        this.doc_.removeEventListener('click', this.onClickOutside_);
+        this.removeIntermediateIframe_();
+      }
+    };
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.CONTAINER;
+  }
+
+  /** @override */
+  buildCallback() {
+    console.log('buildCallback');
+    this.win.addEventListener('message', (event) => {
+      // Make sure the postMessage comes from the iframe origin
+      if (
+        this.validateMessageOrigin_(event.origin) &&
+        this.validateMessageSource_(event.source)
+      ) {
+        this.handleIntermediateIframeMessage_(event.data);
+      }
+    });
+    this.win.addEventListener('load', () => {
```

Comment:
don't make iframes in build callback - this will violate the amp privacy preserving prerender.  Gate the iframe on the page being visible - moving the iframe to the layout callback will do this.

Reply:
done